### ISSUE TITLE
Send Buffer Access

### ISSUE THREAD

**Issue by mpiforumbot** [mpi-forum/mpi-forum-historic#45]
**Originally by _erezh_ on 2008-10-16 15:36:19 -0500**

---
# Remove Send Buffer Access Restrictionworkgroup list [mailto:mpi-buff_access@lists.mpi-forum.org]
## April meeting straw votes (4/29/2008)

Yes: 16[[BR]]
No: 1[[BR]]
Abstain: 11[[BR]]
## Jul meeting straw votes (7/1/2008)

Yes: 22[[BR]]
No: 0[[BR]]
Dont care: 0[[BR]]
Dont have enough info: 4[[BR]]
## Background:

MPI 1.1 standard prohibits users from accessing for read their send buffer until the send operation completes. Be it access in the same thread in case of an async send operation or by another thread in the case of blocking send operation. The rational in the MPI 1.1 standard was to enable the performance for DMA engine that is not cache-coherent with the main processor.
## Proposal:

Remove the read access restriction on the send buffers.
## Rational:

This restriction is counter intuitive for many programmers. It leads to programs that are not compliant with the spec, unaware of this limitation. This limitation prohibits a common usage of the async send API’s, and or the blocking sends.[[BR]]
For example the following code sequence is not compliant with the spec.

```
#!c
void *buffer;
int count,tag;
...
MPI_Isend(buffer, count, MPI_CHAR, 3, tag, MPI_COMM_WORLD, &request1);
MPI_Isend(buffer, count, MPI_CHAR, 4, tag, MPI_COMM_WORLD, &request2);
```

The email thread discussing the cache-coherency machines limitation concluded that it’s a non issue for these machines. The application should resolve the multi-threaded non cache-coherency issues anyhow for sending data.
### Impact on MPI implementations:

This change will render implementations that modify the user buffer in-place (like byte-swap) invalid.
### Impact on MPI applications:
- no impact on applications that are compliant with the existing standard.
- applications that are not compliant with the existing standard and are accessing the send buffer will become compliant (and portable).
- new applications can take advantage of accessing the send buffer to improve performance (rather than copying the send buffer, or using blocking operations).
### Proposed changes to the MPI document:

_The location pointers are refering to the [MPI 2.1 Standard](http://www.mpi-forum.org/docs/mpi21-report.pdf)_

---

_Section 3.4 Page 38 Line 7_[[BR]]
**Change:**[[BR]]
  The send call described in Section 3.2.1 is blocking: it does not return until the message data and envelope have been safely stored away so that the sender is free to ~~access and overwrite~~ the send buffer.

-_To:_\* _(change “access and overwrite” to “modify”)_[[BR]]
  The send call described in Section 3.2.1 is blocking: it does not return until the message data and envelope have been safely stored away so that the sender is free to **modify** the send buffer.

---

_Section 3.4 Page 41 Line 1_[[BR]]
**Change:**[[BR]]
  In a multi-threaded implementation of MPI, the system may de-schedule a thread that is blocked on a send or receive operation, and schedule another thread for execution in the same address space. In such a case it is the user’s responsibility not to ~~access or~~ modify a communication buffer until the communication completes. Otherwise, the outcome of the computation is undefined.

-_To:_\* _(remove “access or”)_[[BR]]
  In a multi-threaded implementation of MPI, the system may de-schedule a thread that is blocked on a send or receive operation, and schedule another thread for execution in the same address space. In such a case it is the user’s responsibility not to modify a communication buffer until the communication completes. Otherwise, the outcome of the computation is undefined.

---

_Section 3.4 Page 41 Lines 7-12_[[BR]]
**Remove:**[[BR]]
  ~~Rationale. We prohibit read accesses to a send buffer while it is being used, even though the send operation is not supposed to alter the content of this buffer. This may seem more stringent than necessary, but the additional restriction causes little loss of functionality and allows better performance on some systems — consider the case where data transfer is done by a DMA engine that is not cache-coherent with the main processor. (End of rationale.)~~

---

_Section 3.9 Page 71 Line 44_[[BR]]
**Change:**[[BR]]
  If the request is for a send with ready mode, then a matching receive should be posted before the call is made. The communication buffer should not be ~~accessed~~ after the call, and until the operation completes.

-_To:_\* _(change “accessed” to “modified”)_[[BR]]
  If the request is for a send with ready mode, then a matching receive should be posted before the call is made. The communication buffer should not be **modified** after the call, and until the operation completes.

---

_Section 5.1 Page 131 Line 11_[[BR]]
**Change:**[[BR]]
  Collective routine calls can (but are not required to) return as soon as their participation in the collective communication is complete. The completion of a call indicates that the caller is now free to ~~access~~ locations in the communication buffer.

-_To:_\* _(change “access” to “modify”)_[[BR]]
  Collective routine calls can (but are not required to) return as soon as their participation in the collective communication is complete. The completion of a call indicates that the caller is now free to **modify** locations in the communication buffer.

---

_Section 11.3 Page 325 Line 20-31_[[BR]]
**Remove:**[[BR]]
  ~~Rationale. The rule above is more lenient than for message passing, where we do not allow two concurrent sends, with overlapping send buffers. Here, we allow two concurrent puts with overlapping send buffers. The reasons for this relaxation are~~
1. ~~Users do not like that restriction, which is not very natural (it prohibits concurrent reads).~~
2. ~~Weakening the rule does not prevent efficient implementation, as far as we know.~~
3. ~~Weakening the rule is important for performance of RMA: we want to associate one synchronization call with as many RMA operations is possible. If puts from overlapping buffers cannot be concurrent, then we need to needlessly add synchronization points in the code.~~[[BR]]
   
   ~~(End of rationale.)~~
## Entry for the Change Log

Section 3.4, page 38, Section 3.9, page 68, and Section 5.1, page 129.  

The read access restriction on the send buffer for blocking, non blocking and collective API has been lifted. It is permitted to access for read the send buffer while the operation is in progress.

**Comment by mpiforumbot** [comment]
**Originally by _erezh_ on 2008-10-16 15:48:54 -0500**

---

## Comments

Comment by Dick Treumann - The suggested change for collective communication is to strong. CC operations have both send and receive buffers. Before return from a CC, any access (not just modify) to a CC receive buffer must remain invalid.  The argument for access before completion applies only to CC send buffers

Reply by Erez Haba - The suggested wording is **less** restrictive and say nothing about accessing the buffers for read. I think that it is obvious that reading receive buffer before the call is complete, results in invalid data. I did not want to add wording to explicitly allow reading the send buffer.

**Comment by mpiforumbot** [comment]
**Originally by _buntinas_ on 2008-12-05 16:03:52 -0600**

---

Darius will review.

**Comment by mpiforumbot** [comment]
**Originally by _asupalov_ on 2008-12-05 16:11:46 -0600**

---

By the way of reviewing this proposal, let's consider the following scenario:
- In the MPI_Isend, MPI maps the send buffer into the address space of the receiving process.
- In the matching MPI_Recv, the receiving process makes a copy of the mapped send buffer into the receive buffer.
- Once the copy is complete, the send buffer is mapped back into the sender address space during the wait/test call.

What will happen one tries to access the send buffer in between?

**Comment by mpiforumbot** [comment]
**Originally by _asupalov_ on 2008-12-05 16:15:59 -0600**

---

Another scenario for review:
- In the MPI_Isend, the MPI swaps the bytes from little-endian (platform) to big-endian (network) order
- The data is transferred in background
- The buffer is restored to its original state in the respective test/wait call, or by a helper thread once the background transfer is complete.

What will happen if this buffer is accessed in between?

**Comment by mpiforumbot** [comment]
**Originally by _asupalov_ on 2008-12-05 16:19:12 -0600**

---

Transforming two notes above into comments (still learning).

I don't think proposal is good for MPI implementations. It restricts the internal MPI working way too much.

Let's consider the following scenario:
- In the MPI_Isend, MPI maps the send buffer into the address space of the receiving process.
- In the matching MPI_Recv, the receiving process makes a copy of the mapped send buffer into the receive buffer.
- Once the copy is complete, the send buffer is mapped back into the sender address space during the wait/test call.

What will happen one tries to access the send buffer in between? 

Another scenario for review: 
- In the MPI_Isend, the MPI swaps the bytes from little-endian (platform) to big-endian (network) order
- The data is transferred in background
- The buffer is restored to its original state in the respective test/wait call, or by a helper thread once the background transfer is complete. 

What will happen if this buffer is accessed in between?

**Comment by mpiforumbot** [comment]
**Originally by _rlgraham_ on 2008-12-13 17:23:23 -0600**

---

Reviewed as chapter author for chapter 3.  Changes ok - Rich Graham

**Comment by mpiforumbot** [comment]
**Originally by _traff_ on 2008-12-14 19:39:15 -0600**

---

Textually ok (but I share some of the doubts about the contents of this proposal),
except for the collectives part, where I'd suggest either not to change the text
(except that "communication buffer", p.131, line 13, should be "communication buffers"), or extend the explanation.

Jesper

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2008-12-15 18:39:50 -0600**

---

Please can you add at the beginning of the Change-Log:

Section 3.4 on page 38, Section 3.9 on page 68, and Section 5.1 on page 129.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-01-09 06:46:31 -0600**

---

Sorry, that I did not found earlier, i.e., before 2nd vote.
Please add the following modification of MPI-2.1:

-Section 3.7 Page 52 Lines 5-6_[[BR]]
*_Change:**[[BR]]
  The sender should not ~~access~~ any part of the send buffer after a nonblocking
  send operation is called, until the send completes.

-_To:_\* _(change “access” to “modify”)_[[BR]]
  The sender should not **modify** any part of the send buffer after a nonblocking
  send operation is called, until the send completes.

Add **Section 3.7.2, page 49,** to the Change-Log Entry.

**Comment by mpiforumbot** [comment]
**Originally by _asupalov_ on 2009-04-03 04:31:08 -0500**

---

Replying to RolfRabenseifner:

> Sorry, that I did not found earlier, i.e., before 2nd vote.
> Please add the following modification of MPI-2.1:

Technically, this should be a separate ticket.

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2009-07-10 15:19:21 -0500**

---

The collectives chapter changes are complete and checked into svn for this ticket.  I've removed the pdf, since it may have been incomplete (I only implemented the collective chapter changes).  The page in the current pdf is MPI page 131 (absolute page 157 in the document) and the ChangeLog entry which shows up on page 560 (absolute 586).  These page numbers may change though if other tickets move text around.

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2009-07-10 15:29:11 -0500**

---

**Attachment added:** [`chap-coll.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/45/chap-coll.pdf) (41.9 KiB)
Whoops, I guess I can't delete the file, so I've uploaded a new one to overwrite it -- this only shows the collectives chapter changes.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-07-13 08:30:43 -0500**

---

**Attachment added:** [`change-log_ticket45_item5.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/45/change-log_ticket45_item5.pdf) (85.0 KiB)

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-07-13 10:37:46 -0500**

---

PDF-Review of the part in Chap. 5 (coll): Okay

**Comment by mpiforumbot** [comment]
**Originally by _traff_ on 2009-07-13 10:44:37 -0500**

---

Change in one-sided chapter (removal of "rationale") has been carried out, reviewers
please check (full text of Chapter 11 attached)

**Comment by mpiforumbot** [comment]
**Originally by _traff_ on 2009-07-13 10:45:25 -0500**

---

**Attachment added:** [`onesidedtickets.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/45/onesidedtickets.pdf) (312.3 KiB)

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-07-13 12:33:13 -0500**

---

PDF-Review of the part in Chap. 11 (rma): Okay

**Comment by mpiforumbot** [comment]
**Originally by _rlgraham_ on 2009-08-09 21:10:36 -0500**

---

changes made to chapter 3

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-08-20 06:23:06 -0500**

---

**Attachment added:** [`ticket45_all.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/45/ticket45_all.pdf) (504.9 KiB)

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-08-20 06:25:11 -0500**

---

All changes are done, therefore switched to "Waiting for PDF reviews".

First review: all okay.

**Comment by mpiforumbot** [comment]
**Originally by _erezh_ on 2009-08-21 12:40:12 -0500**

---

reviewed, all okay.

**Comment by mpiforumbot** [comment]
**Originally by _moody20_ on 2009-08-31 17:53:37 -0500**

---

Reviewed ticket45_all.pdf: OK.

**Comment by mpiforumbot** [comment]
**Originally by _gropp_ on 2009-09-02 13:58:54 -0500**

---

Reviewed master copy and ok.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2010-09-18 05:01:18 -0500**

---

This ticket is (long-since) complete; marking it resolved/text committed.
